src/cluster.h

```diff
@@ -173,257 +51,102 @@ typedef struct {
-#define CLUSTERMSG_FLAG0_FORCEACK (1<<1) /* Give ACK to AUTH_REQUEST even if
-                                            master is up. */
-#define CLUSTERMSG_FLAG0_EXT_DATA (1<<2) /* Message contains extension data */
-
 /* ---------------------- API exported outside cluster.c -------------------- */
+/*
+ * Functions requiring per-clustering mechanism implementation: Lifecycle events
+ */
 void clusterInit(void);
 void clusterInitListeners(void);
 void clusterCron(void);
 void clusterBeforeSleep(void);
-clusterNode *getNodeByQuery(client *c, struct redisCommand *cmd, robj **argv, int argc, int *hashslot, int *ask);
-int verifyClusterNodeId(const char *name, int length);
-clusterNode *clusterLookupNode(const char *name, int length);
-int clusterRedirectBlockedClientIfNeeded(client *c);
-void clusterRedirectClient(client *c, clusterNode *n, int hashslot, int error_code);
-void migrateCloseTimedoutSockets(void);
 int verifyClusterConfigWithData(void);
+void clusterUpdateMyselfFlags(void);
 unsigned long getClusterConnectionsCount(void);
+void clusterUpdateMyselfHostname(void);
+void clusterUpdateMyselfAnnouncedPorts(void);
+void clusterUpdateMyselfIp(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: cluster commands
+ */
+void clusterCommand(client *c);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Pub/Sub and module messages
+ */
 int clusterSendModuleMessageToTarget(const char *target, uint64_t module_id, uint8_t type, const char *payload, uint32_t len);
 void clusterPropagatePublish(robj *channel, robj *message, int sharded);
+void slotToChannelAdd(sds channel);
+void slotToChannelDel(sds channel);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Debug and info actions
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
+sds clusterGenNodesDescription(client *c, int filter, int use_pport);
+sds genClusterInfoString(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: "Accessor" functions
+ */
+char* clusterNodeLastKnownIp(clusterNode *node);
+int clusterNodePort(clusterNode *node);
+clusterNode* clusterNodeGetSlaveof(clusterNode *node);
+int clusterNodeConfirmedReachable(clusterNode* node);
+int clusterNodeIsMaster(clusterNode* node);
+int clusterNodeIsSlave(clusterNode* node);
+int clusterNodeIsFailing(clusterNode* node);
+int clusterNodeTimedOut(clusterNode* node);
+int clusterNodeIsMyself(clusterNode* node);
+int clusterNodeIsNoFailover(clusterNode* node);
+clusterNode* getMyClusterNode(void);
+clusterNode* getNodeBySlot(int slot);
+clusterNode* getMigratingSlotDest(int slot);
+clusterNode* getImportingSlotSource(int slot);
+int isClusterHealthy(void);
+uint16_t getClusterNodeRedirectPort(clusterNode* node, int use_pport);
+const char *getPreferredEndpoint(clusterNode *n);
+clusterNode *clusterLookupNode(const char *name, int length);
+int isClusterManualFailoverInProgress(void);
```

Comment:
This feels wrong in the API. I would expect Redis to be driving this, because it knows the detail of the state. I would expect both clustering systems to "initiate" the failover, but they shouldn't be the one monitoring it.

Reply:
Thought about this more, and yeah I still think there should be a way to call "start write pause" and "stop write pause" and then get ride of this special manual failover state for cluster. I think this is just leftover code, I could send a PR for this now I suppose. Either than that, or we should expose an API like, "Stepdown as soon as you can", and the master will wait for a replica to sync and step down.

Reply:
It does feel wrong but it's required in replication.c

Reply:
Ok, to make my position clear, I would like us to build a coherent API, even if that requires some modification to how the code is currently is implemented. If you want to split between refactoring and rewriting, I would prefer we do the rewrite first then the refactor.

Reply:
Sorry for the crossed wires but my response was to your initial comment here, not the second one. I think I did not refresh my browser or something when I wrote it. Anyway, your proposal does sound like a good idea though I would prefer to keep this PR focused on refactoring.

Reply:
Cool, we had a brief talk with the core team and agreed that early in Redis 8 was the time for refactoring, so we should be able to get all this stuff done.